# iDog2.0 Data Processing Pipelines

This project is designed to create high-quality, consistent, and reproducible data for iDog. We have developed data processing pipelines for major assay types generated by the project: bulkRNA-seq, ATAC-seq, and BS-seq.

## Features
In recent years, we have substantially enhanced iDog, culminating in the release of iDog 2.0 based on newly reference CanFam4. As the data portal of the international Dog10K project and consortium, this updated version encompasses data generated by the Dog10K project, along with a wealth of data from numerous other projects published in recent years. It conducts comprehensive and integrated analyses encompassing genome, transcriptome, epigenome, phenome, and single-cell transcriptome data, spanning a vast array of samples ranging from ancient to modern canines, diverse breeds, varying ages, and different tissues or cell types of canines. This establishes a basis for thorough analysis of the Canidae genetic variants and function, tissue transcriptional expression and regulation in normal and disease states, and neuronal cells and circuits within neurological evolution. These investigations notably advance further studies on domestication, traits, behavior, morphology, and disease susceptibility.


## Dependencies

Before running the script, ensure that you have the following dependencies installed:
- Python (3.x)
  * pandas
  * intervaltree
  * pigz
  * glob
  * argparse
- R (4.1.1+)
  * DESeq2
  * pheatmap
  * RColorBrewer
  * ggplot2
  * cowplot
  * hexbin
  * ggrepel
  * ggsignif
  * vsn
- fastp (0.23.2+)
- bowte2 (2.5.3+)
- samtools (1.15.1+)
- picard (2.22.8+)
- bedtools (2.31.1+)
- macs2 (2.2.4+)
- bismark (0.23.1+)
- kallisto (0.46.0+)
- STAR (2.73a)
- rsem (1.3.3+)
- sratoolkit (3.0.2)
- pigz (2.4+)

## Installation

```bash
git clone https://github.com/Br1anChou/idog.git
```

## Usage

#### Decompression(optional)
```bash
vdb-validate sample_id.sra

fasterq-dump -e ncpus --split-3 --outdir /PATH/sample_id/reads sample_id.sra
```

#### bulkRNA-seq
```bash
bash bulkRNA/bulkRNA-seq_E4.sh <fastq_dir> <out_dir> <sample_id> <ncpus> <ramGB>
```

#### ATAC-seq
```bash
bash ATAC/align_E4.sh <fastq_dir> <out_dir> <sample_id> <ncpus> <ramGB>

python ATAC/poolTagAligns.py <output_name> <input_file1> <input_file2> ...

bash ATAC/callPeaks_ATAC_E4.sh <in_file> <out_dir> <sample_id>
```

#### BS-seq
```bash
bash WGBS/WGBS_E3.sh <fastq_dir> <out_dir> <sample_id> <ncpus> <ramGB>
```

## Cition
Yanhu Liu, Yibo Wang, Jiani Sun, Demian Kong, Bowen Zhou, Mengting Ding, Yuyan Meng, Guangya Duan, Ying Cui, Zhuojing Fan, Ya-Ping Zhang, Wenming Zhao, Bixia Tang, iDog: a multi-omics resource for canids study, Nucleic Acids Research, 2024;, gkae1031, https://doi.org/10.1093/nar/gkae1031
